new feature: plugin modal view needed for quicker edit

shamsbd71 opened the pull request:
in admin just the way joomla has module modal edit view, we need to have plugin edit view in modal to save its settings. we dont want to forward users to find and edit things
## Test Instructions:

i have showed in com_redirect with full functional language, helper and view files. so if you visit com_redirect
administrator/index.php?option=com_redirect
you will see how its working.

![image](https://cloud.githubusercontent.com/assets/1936565/8352947/1f004782-1b5b-11e5-8e9b-9c5ce3f7530b.png)
thanks

shamsbd71 committed:
joomla plugin edit view in modal

for quick user action to modify modules settings in popup a modal view is necessary to save it.

shamsbd71 committed:
Merge pull request #1 from shamsbd71/shamsbd71-j3-patch-pluginmodal

New feature: joomla plugin edit view in modal

n9iels commented:
Can you please give some test instructions? And also tell us what and where you exactly add new features?

shamsbd71 committed:
after save the plugin on modal ,close modal

after save the modal window for plugin edit, we need to close it as its main work done, we dont need to show plugin lists. so close the modal

shamsbd71 commented:
hi @n9iels 
i'm building a Joomla! eCommerce extension for digital goods, its been called as "Digicom". feel free to check at https://github.com/themexpert/digicom

we have plugins for payment and some other operations. and users need to fill the info by themselves.
but to do that they need to do many clicks, search , edit , save and then come back to the extension 
:(
so to avoid this situation we want to give them links at sidebar (that we have created) and users will click on it, it opens, they save and its gone :)

i'm attaching some screen capture, hope it will help
thanks
![image](https://cloud.githubusercontent.com/assets/1936565/8348543/562c4370-1b36-11e5-99e7-63e3ad2cff37.png)

![image](https://cloud.githubusercontent.com/assets/1936565/8348579/b227c7a8-1b36-11e5-9403-c749bd70f8c9.png)

shamsbd71 commented:
One more thing, please check this file
https://github.com/themexpert/digicom/blob/develop/plugins/digicom_pay/paypal/paypal.php
you will see a method name : onSidebarMenuItem
this is the event i have created for #digicom to show the links from plugin

** joomla already have the same feature for modules. and we can live edit the modules from menu item edit view. we need that for module too. :)
let me know if i need to provide something more.

Fedik commented:
seems it could be a nice feature in your extension, 
but I have doubt where such thing can be useful in Joomla ... I did not understand this part from your explanation

shamsbd71 commented:
@Fedik 
you are right that in general Joomla! doesn't need plugin to edit on the fly like module as plugin has no option to load on specific pages. but
as plugin is obvious and many Joomla! extensions require to get info/settings to be updated by admin.
so if you enable this options as built in, from now many developer will use it to give the Joomla! users a nice n quicker action to edit plugin the way we can do with modules.

and it will improve users experience
thanks

brianteeman commented:
I "think" I can see the use case for this directly in the core - if I
understand correctly

For example in com_redirect by default you get a message
`The Redirect Plugin is disabled. Enable it in the Plugin Manager.`

If I understand you correctly then with this PR that message could be
changed to a modal link to the plugin?



On 25 June 2015 at 09:30, Abu-Huraira <notifications@github.com> wrote:

> @Fedik <https://github.com/Fedik>
> you are right that in general Joomla! doesn't need plugin to edit on the
> fly like module as plugin has no option to load on specific pages. but
> as plugin is obvious and many Joomla! extensions require to get
> info/settings to be updated by admin.
> so if you enable this options as built in, from now many developer will
> use it to give the Joomla! users a nice n quicker action to edit plugin the
> way we can do with modules.
>
> and it will improve users experience
> thanks
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259#issuecomment-115161599>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

shamsbd71 commented:
@brianteeman 
you are right, it can be changed to modal so users can change it directly from that view without visiting anymore pages.

codexpert commented:
+1 @brianteeman also from 'com_finder' we can enable/disable the plugin.

There is many use case actually. Many components depends on lot of plugin, for instance 'Virtuemart, Payplans, Jomsocial' etc. So for every change we need to go to plugin manager > search plugin > open and change settings > back to component again.

brianteeman commented:
Can I suggest that you update the pull request to include com_redirect.
Then this can be used as an example AND makes testing easier.

The easier it is to test the more likely it is to be reviewed
On 25 Jun 2015 10:35, "Parvez Akther" <notifications@github.com> wrote:

> +1 @brianteeman <https://github.com/brianteeman> also from 'com_finder'
> we can enable/disable the plugin.
>
> There is many use case actually. Many components depends on lot of plugin,
> for instance 'Virtuemart, Payplans, Jomsocial' etc. So for every change we
> need to go to plugin manager > search plugin > open and change settings >
> back to component again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259#issuecomment-115185911>.
>

shamsbd71 commented:
thank you @brianteeman. will do that

shamsbd71 committed:
getRedirectPluginId method added

getRedirectPluginId to get the system redirect plugin id for quicker action

shamsbd71 committed:
com_redirect language updated

1 language modified and one added for com_redirect to match edit plugin message

shamsbd71 committed:
get the Redirect plugin id 

get the redirectpluginid value from RedirectHelper

shamsbd71 committed:
modal edit view for system Redirect Plugin

onclick it will show plugin edit view in modal for system redirect plugin

shamsbd71 committed:
btn class added to edit plugin link

shamsbd71 commented:
I have updated the PR, now it will show in com_redirect in this way : 
![image](https://cloud.githubusercontent.com/assets/1936565/8352947/1f004782-1b5b-11e5-8e9b-9c5ce3f7530b.png)

* test instruction updated in main comment

brianteeman commented:
Thanks

On 25 June 2015 at 12:01, Abu-Huraira <notifications@github.com> wrote:

> I have updated the PR, now it will show in com_redirect in this way :
> [image: image]
> <https://cloud.githubusercontent.com/assets/1936565/8352947/1f004782-1b5b-11e5-8e9b-9c5ce3f7530b.png>
>
>    - test instruction updated in main comment
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259#issuecomment-115210873>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

Bakual commented on administrator/components/com_plugins/views/plugin/tmpl/edit.php:
```diff
@@ -20,6 +20,10 @@
 			Joomla.submitform(task, document.getElementById('style-form'));
+			
+			if (self != top){
+				window.top.setTimeout('window.parent.jModalClose()', 1000);
```
It may be obvious but I suck at JS. What's the purpose of this?

shamsbd71 replied:
when we save the plugin it will show the plugin list as its loaded in iframe.
btw, module modal used the same approach and more, i just picked up the code necessary to close modal after it save

Bakual replied:
Thanks

Bakual replied:
Hmm, can't see that in https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_modules/views/module/tmpl/modal.php. Maybe it got changed recently?

brianteeman replied:
Here I think #6804

On 25 June 2015 at 12:55, Thomas Hunziker <notifications@github.com> wrote:

> In administrator/components/com_plugins/views/plugin/tmpl/edit.php
> <https://github.com/joomla/joomla-cms/pull/7259#discussion_r33246541>:
>
> > @@ -20,6 +20,10 @@
> >  	{
> >  		if (task == 'plugin.cancel' || document.formvalidator.isValid(document.getElementById('style-form'))) {
> >  			Joomla.submitform(task, document.getElementById('style-form'));
> > +			
> > +			if (self != top){
> > +				window.top.setTimeout('window.parent.jModalClose()', 1000);
>
> Hmm, can't see that in
> https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_modules/views/module/tmpl/modal.php.
> Maybe it got changed recently?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259/files#r33246541>.
>



-- 
Brian Teeman
Co-founder Joomla! and OpenSourceMatters Inc.
http://brian.teeman.net/

shamsbd71 replied:
sorry i didn't mention the exact file
https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_modules/views/module/tmpl/edit.php#L123

but seems its already changed to jquery

Fedik commented:
If we want introduce new "popup", it already should be without mootools dependency

Bakual commented on administrator/components/com_redirect/views/links/view.html.php:
```diff
@@ -129,7 +133,7 @@ protected function addToolbar()
 		}
 
-		if ($canDo->get('core.admin') || $canDo->get('core.options'))
+		if ($canDo->get('core.admin'))
```
Revert that line to the original. I guess it comes from a bad merge.

Bakual commented:
Some things to think about: You create a new method in the RedirectHelper to fetch the ID.
I think that could be something which could be useful for other extensions as well if we introduce this modal. So I guess we could add a `getPluginId()` method to JPluginHelper instead.
The method also shouldn't use JError (it's deprecated). I'm not even sure if we need to do something with the error or if we just want to return `false`.

By the way, the `RedirectHelper::isEnabled()` could probably be replaced with `JPluginHelper::isEnabled('system', 'redirect')`, but that's another topic.

The modal imho should be wider. Similar to the module modal loaded from the menu item.

Maybe the button also should only be there if we have found a plugin id. Otherwise you get a link to the list.

Bakual commented:
`administrator/components/com_redirect/views/links/tmpl/default.php` has a lot of unrelated changes. You probably had an outdated file there. Make sure you work from current staging code and only change what you want.

dgt41 commented on administrator/components/com_redirect/views/links/tmpl/default.php:
```diff
@@ -15,6 +15,7 @@
 JHtml::_('bootstrap.tooltip');
 JHtml::_('behavior.multiselect');
 JHtml::_('formbehavior.chosen', 'select');
+JHTML::_('behavior.modal');
```
Can we directly introduce a bootstrap modal here? It  seems like we stepping backwards introducing a mootools dependency here

shamsbd71 replied:
@Bakual , @dgt41 
thank you, i have implemented the bootstrap way

dgt41 commented on administrator/components/com_redirect/views/links/tmpl/default.php:
```diff
@@ -44,22 +45,31 @@
+				<?php echo JText::_('COM_REDIRECT_PLUGIN_DISABLED'); ?>
 		<?php endif; ?>
+
+			<a class="modal btn btn-small" rel="{handler: 'iframe', size: {x: 900, y: 550}}" 
```
you can use the bootstrap alternative her, something like:
```php
			<?php echo JHtml::_(
					'bootstrap.renderModal',
					'module' . $module->id . 'Modal',
					array(
						'url' => $link,
						'title' => JText::_('COM_MENUS_EDIT_MODULE_SETTINGS'),
						'height' => '300px',
						'width' => '800px',
						'footer' => '<button class="btn" data-dismiss="modal" aria-hidden="true">'
							. JText::_("JLIB_HTML_BEHAVIOR_CLOSE") . '</button>'
							. '<button class="btn btn-success" data-dismiss="modal" aria-hidden="true" onclick="jQuery(\'#module' . $module->id . 'Modal iframe\').contents().find(\'#saveBtn\').click();">'
							. JText::_("JSAVE") . '</button>'
					)
				); ?>
```

dgt41 commented on administrator/components/com_plugins/views/plugin/tmpl/modal.php:
```diff
@@ -0,0 +1,26 @@
+		<?php echo JText::_('JCANCEL');?></button>
+	</div>
+	<div class="clearfix"></div>
+</div>
```
If you use the bootstrap modal these buttons can go to the modal footer

shamsbd71 replied:
thanks, i just did that

shamsbd71 committed:
bootstrapization of modal view

following this file https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_modules/views/module/tmpl/modal.php

shamsbd71 committed:
bootstrapization of plugin edit view modal close

followed by https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_modules/views/module/tmpl/edit.php#L123

shamsbd71 committed:
bootstrapization of modal method call

followed by: https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_menus/views/item/tmpl/edit_modules.php#L102

shamsbd71 committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into staging

marged with joomla latest staging 3.4.3

shamsbd71 committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into staging

shamsbd71 commented:
am i missing anything?

shamsbd71 committed:
JPluginHelper::getPlugin now return id #7355

shamsbd71 committed:
JPluginHelper::protected static function load() updated to load matched plugin regardless its enabled or not, as we have isEnabled() method to test it

shamsbd71 committed:
JPluginHelper::loadPlugin handled need only enabled as thats used saperately as object. so keeping it

shamsbd71 committed:
redirect links view updated to use jpluginhelper code and get id from redirect helper

shamsbd71 commented:
as i get request to check if i can use "JPluginHelper::isEnabled('system', 'redirect')"

i have used it, and to get the plugin id from JPluginHelper::getPlugin('system', 'redirect'); but it will only return when the plugin is enabled.

so i had to keep it in RedirectHelper

if you i should add in JPluginHelper a method to get only id, then i can do that too. 

thanks

shamsbd71 committed:
modal wrapper moved from alert wrapper and modal title language added

shamsbd71 committed:
Merge branch 'staging' of https://github.com/joomla/joomla-cms into staging

brianteeman committed:
codestyle changes

brianteeman commented:
I submitted a PR to your repo to fix the codestyle change. If you apply it then it should pass travis (the codechecker)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="http://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/joomla-cms/7259</a>.</sub>

shamsbd71 committed:
Merge pull request #2 from brianteeman/patch-1

codestyle changes

shamsbd71 committed:
Merge branch 'staging' of github.com:shamsbd71/joomla-cms into staging

shamsbd71 committed:
pull latest commits from current joomla staging

shamsbd71 committed:
batch permission n modal marged with joomla staging: links list in admin

shamsbd71 commented:
is there anyone can tell me what to do with this issue or it should be closed ?

brianteeman commented:
@dgt41 @andrepereiradasilva Can you guys take a look at this please. Seems useful to me but need more experienced people than me to make sure the issues raised above have been addressed. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/joomla-cms/7259</a>.</sub>

AnishaVora commented:
I have tested this item :white_check_mark: successfully on 4a1724b7b8a3cdb3620a3e1b8caa6069cf7511fd<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/joomla-cms/7259</a>.</sub>

gunjanpatel commented:
Setting to `pending` as we have code.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/joomla-cms/7259</a>.</sub>

florian1995 commented:
I have tested this item :white_check_mark: successfully on 4a1724b7b8a3cdb3620a3e1b8caa6069cf7511fd<br /><br />I tested it with the given component and plugin, and it works perfectly.
I get the edit button and can enable and disable the plugin, so i checked it in the joomla backend and it worked.
Tested@icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/joomla-cms/7259</a>.</sub>

roland-d commented:
@shamsbd71 Thank you for this feature. I do find it useful as well and see how it can be useful in 3PD extensions.

One question I have is why use a modal popup instead of just have a button that enables the plugin directly? That would save a few clicks.

This would be needed for Smart Search as well.

brianteeman commented:
@roland-d if there are plugin options it would be better to show them to the user before enabling it. My 2c

roland-d commented:
@brianteeman Good point, something I was thinking as well after posting my comment. So let's stick to the modals. I do like the original idea.

@shamsbd71 Could you fix the merge conflicts please?

brianteeman commented:
@shamsbd71 are you able to resolve the merge conflicts and then we can get this merged<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/tracker/joomla-cms/7259</a>.</sub>

franz-wohlkoenig commented:
only PR-dev can resolve merge Conflicts?

franz-wohlkoenig commented:
who can resolve merge Conflicts?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/tracker/joomla-cms/7259</a>.</sub>

brianteeman commented:
@franz-wohlkoenig anyone with commit access can do it

franz-wohlkoenig commented:
thanks for Info, @brianteeman is anyone having commit Access willing to resolve merge Conflicts?

roland-d commented:
@wilsonge I think it is useful to take this into Joomla 4, what do you think?

brianteeman commented:
> @wilsonge I think it is useful to take this into Joomla 4, what do you think?
 
Why wait for J4 - joomla 3 has over 2 years of life

roland-d commented:
because there are no new feature releases planned at this moment with exception for 3.8 being the router release.

brianteeman commented:
@roland-d where is that documented - its not what the roadmap says https://developer.joomla.org/roadmap.html

roland-d commented:
We read the roadmap differently then. It says the following:
```
The following are the primary features introduced to Joomla! 3.8.

    Modern Routing System
    Joomla! 4 compatibility layer
```

The features in development are set as the Modern Routing System and General improvements.

brianteeman commented:
Keep reading further
> Proposed Features

> The following are suggested improvements which would be highly favored for this release. If you would like to assist or contribute to them, please contact the release lead or release team for more information.

> ** To be determined**

roland-d commented:
In that case @mbabker do you want this in 3.8? If so we probably need to find someone who wants to clean this up and propose a new PR or for @shamsbd71 to fix this.

mbabker commented:
Personally my aim is to keep the major focus on the router.  As seen on the
project boards, I've flagged a bunch of PRs for 3.8 that are relatively
smaller in nature.  So honestly, it's going to be a case by case basis
right now.

On Sat, May 13, 2017 at 3:51 PM RolandD <notifications@github.com> wrote:

> In that case @mbabker <https://github.com/mbabker> do you want this in
> 3.8? If so we probably need to find someone who wants to clean this up and
> propose a new PR or for @shamsbd71 <https://github.com/shamsbd71> to fix
> this.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259#issuecomment-301273992>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWfoYZXuNwlJJ5_z5UTy8OnTB2wsID5ks5r5he4gaJpZM4FKrT8>
> .
>
-- 
- Michael Please pardon any errors, this message was sent from my iPhone.

rdeutz commented:
So 3.8 or 4.0 @mbabker @wilsonge who will take it :-)

brianteeman commented:
3.8 please

On 24 May 2017 10:52 a.m., "Robert Deutz" <notifications@github.com> wrote:

> So 3.8 or 4.0 @mbabker <https://github.com/mbabker> @wilsonge
> <https://github.com/wilsonge> who will take it :-)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259#issuecomment-303675126>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABPH8VF09esJ3WFUZm6dFpNO8p_KmHyOks5r8_3UgaJpZM4FKrT8>
> .
>

mbabker commented:
If this is tested and reviewable I've got no issue looking at it for 3.8.  I'm just not making promises on things being included into the release is all.

franz-wohlkoenig commented:
@mbabker PR have 2 successfully Test, but conflicts. Close PR?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/tracker/joomla-cms/7259</a>.</sub>

mbabker commented:
Closing the PR just because it has merge conflicts isn't right.  If it's been abandoned though, which seems to be the case, and nobody wants to pick it up to get it updated then yes, it should be closed.

franz-wohlkoenig commented:
Thanks for Answer, @mbabker – if this PR get no Response, it will be closed at 23th July 2017.

brianteeman commented:
I will try to find time to redo this

franz-wohlkoenig commented:
thanks @brianteeman

brianteeman commented:
sorry i took a look but its not a simple case of resolving merge conflicts as the base code has changed considerably since this Pr was made. I still think its a good addition but not something I can do

roland-d commented:
I will have a look

brianteeman commented:
Thanks @roland-d - its all the changes in administrator/components/com_redirect/views/links/tmpl/default.php
that I struggled with

codexpert commented:
@abu would you please check your pr when you back?

On Fri, Jun 23, 2017 at 5:05 PM Brian Teeman <notifications@github.com>
wrote:

> Thanks @roland-d <https://github.com/roland-d> - its all the changes in
> administrator/components/com_redirect/views/links/tmpl/default.php
> that I struggled with
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/joomla/joomla-cms/pull/7259#issuecomment-310637943>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAxSE58x1ica4Lw19JWr1j5Ipm-tWfrEks5sG5vpgaJpZM4FKrT8>
> .
>
-- 

-----
Sent from my iPhone

roland-d commented:
Closing this in favor of #16844 <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/7259">issues.joomla.org/tracker/joomla-cms/7259</a>.</sub>
